README.md

```diff
@@ -1,2 +1,238 @@
+<script>
+  hclust.clusterData(...);
+  hclust.euclideanDistance(...);
+  hclust.avgDistance(...);
+  hclust.minDistance(...);
+  hclust.maxDistance(...);
+</script>
+```
+
+#### Node
+
+`npm install git+https://github.com/greenelab/hclust.git`
+
+or
+
+`yarn add git+https://github.com/greenelab/hclust.git`
+
+then 
+
+```javascript
+import { clusterData } from 'hclust';
+import { euclideanDistance } from 'hclust';
+import { avgDistance } from 'hclust';
+import { minDistance } from 'hclust';
+import { maxDistance } from 'hclust';
+```
+
+---
+
+### `clusterData({ data, key, distance, linkage, onProgress })`
+
+#### Parameters
+
+**`data`**
+
+The data you want to cluster, in the format:
+
+```javascript
+[
+  ...
+  [ ... 1, 2, 3 ...],
+  [ ... 1, 2, 3 ...],
+  [ ... 1, 2, 3 ...],
+  ...
+]
+```
+
+or if `key` parameter is specified:
+
+```javascript
+[
+  ...
+  { someKey: [ ... 1, 2, 3 ...] },
+  { someKey: [ ... 1, 2, 3 ...] },
+  { someKey: [ ... 1, 2, 3 ...] },
+  ...
+]
+```
+
+The entries in the outer array can be considered the `rows` and the entries within each `row` array can be considered the `cols`.
+Each `row` should have the same number of `cols`.
+
+*Default value:* `[]`
+
+**`key`**
+
+A `string` key to specify which values to extract from the `data` array.
+If omitted, `data` is assumed to be an array of arrays.
+If specified, `data` is assumed to be array of objects, each with a key that contains the values for that `row`.
+
+*Default value:* `''`
+
+**`distance`**
+
+A function to calculate the distance between two equal-dimension vectors, used in calculating the distance matrix, in the format:
+
+```javascript
+function (arrayA, arrayB) { return someNumber; }
+```
+
+The function receives two equal-length arrays of numbers.
+
+*Default value:* `euclideanDistance` from this `hclust` package
+
+**`linkage`**
+
+A function to calculate the distance between pairs of clusters, used in determining linkage criterion, in the format:
+
+```javascript
+function (arrayA, arrayB, distanceMatrix) { return someNumber; }
+```
+
+The function receives two sets of indexes and the distance matrix computed between each datum and every other datum.
+
+*Default value:* `averageDistance` from this `hclust` package  
+*Other built-in values:* `minDistance` and `maxDistance` from this `hclust` package
+
+**`onProgress`**
+
+A function that is called several times throughout clustering, and is provided the current progress through the clustering, in the format:
+
+```javascript
+function (progress) { }
+```
+
+The function receives the percent progress between `0` and `1`.
+
+*Default value:* an internal function that `console.log`'s the progress
+
+**Note:** [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage) is called in the same places as `onProgress`, if the script is running as a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers).
+
+#### Returns
+
```

Comment:
How does this get returned? These are the keys of the resulting object? I didn't see a point above where you were indicating that folks should assign the result of the function to something. I might have missed it.

Reply:
`clusters`, `distances`, etc are keys of the returned object, yes. It should be clear for people accustomed to javascript.

They can also conveniently destructure the keys like this:
```javascript
const { clusters, distances, order, clustersGivenK } = clusterData({ data, key, distance, linkage, onProgress })

console.log(clusters[0]);
console.log(distances.length);
```